File: 1331-rank-transform-of-an-array/1331-rank-transform-of-an-array.java
class Solution {
    public int[] arrayRankTransform(int[] arr) {
        Map<Integer, Integer> rankMap = new HashMap<>();
        int[] sortedUniqueArray = Arrays.stream(arr).distinct().sorted().toArray();

        int length = sortedUniqueArray.length;
        for (int i = 0; i < length; ++i) 
            rankMap.put(sortedUniqueArray[i], i + 1);

        for (int i = 0; i < arr.length; ++i) 
            arr[i] = rankMap.get(arr[i]);

        return arr;
    }
}

File: 1331-rank-transform-of-an-array/1331-rank-transform-of-an-array.cpp
class Solution {
public:
    vector<int> arrayRankTransform(vector<int>& arr) {
        int length = arr.size();
        vector<int> tempArray(arr);
        sort(tempArray.begin(), tempArray.end());

        map<int, int> rankMap;
        int rank = 1;
        for (int num : tempArray) {
            if (rankMap.find(num) == rankMap.end()) {
                rankMap[num] = rank++;
            }
        }

        vector<int> res(length);
        for (int i = 0; i < length; ++i) {
            res[i] = rankMap[arr[i]];
        }

        return res;
    }
};